### PR TITLE
Pushkarev Ilya \ Lab2 \ Var 2

### PR DESCRIPTION
**Pull request by oleg212**

### CODE REVIEW

**Comment by oleg212** [comment]
some more tests are to be added

**Comment by oleg212** [comment]
done

File: llvm/labs/lab2/pushkarev_ilya/InlineFunc.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  llvm::PreservedAnalyses run(llvm::Function &function,
+                              llvm::FunctionAnalysisManager &) {
+    bool modified = false;
+    std::vector<llvm::Instruction *> instructionsToDelete;
+    llvm::LLVMContext &context = function.getContext();
+
+    for (llvm::BasicBlock &block : function) {
+      for (llvm::Instruction &instruction : block) {
+
+        llvm::CallInst *callInst = llvm::dyn_cast<llvm::CallInst>(&instruction);
```
Comment (m-ly4) [inline llvm/labs/lab2/pushkarev_ilya/InlineFunc.cpp]:
```suggestion
        auto *callInst = llvm::dyn_cast<llvm::CallInst>(&instruction);
```

File: llvm/labs/lab2/pushkarev_ilya/InlineFunc.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        for (llvm::BasicBlock &inlinedBlock : *calledFunction) {
+          llvm::BasicBlock *newBlock =
+              llvm::BasicBlock::Create(context, "", &function, nextBlock);
+          blockMap[&inlinedBlock] = newBlock;
+          nextBlock = newBlock->getNextNode();
+        }
+
+        for (llvm::BasicBlock &inlinedBlock : *calledFunction) {
+          for (llvm::Instruction &inlinedInstruction : inlinedBlock) {
+            llvm::BasicBlock *currentBlock = blockMap[&inlinedBlock];
```
Comment (m-ly4) [inline llvm/labs/lab2/pushkarev_ilya/InlineFunc.cpp]:
```suggestion
          llvm::BasicBlock *currentBlock = blockMap[&inlinedBlock];
          for (llvm::Instruction &inlinedInstruction : inlinedBlock) {
```

File: llvm/labs/lab2/pushkarev_ilya/InlineFunc.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+              llvm::Instruction *newInstruction = inlinedInstruction.clone();
+              if (llvm::isa<llvm::BranchInst>(*newInstruction)) {
+                for (unsigned i = 0; i < newInstruction->getNumOperands();
+                     ++i) {
+                  llvm::Value *operand = newInstruction->getOperand(i);
+                  if (llvm::isa<llvm::BasicBlock>(*operand)) {
+                    newInstruction->setOperand(
+                        i, llvm::dyn_cast<llvm::Value>(
+                               blockMap[llvm::dyn_cast<llvm::BasicBlock>(
+                                   operand)]));
```
Comment (m-ly4) [inline llvm/labs/lab2/pushkarev_ilya/InlineFunc.cpp]:
```suggestion
                  if (auto *bb = llvm::dyn_cast<llvm::BasicBlock>(*operand)) {
                    newInstruction->setOperand(
                        i, llvm::dyn_cast<llvm::Value>(blockMap[bb]));
```

File: llvm/test/lab2/pushkarev_ilya/test.ll [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+; CHECK-NEXT:  %6 = load i32, ptr %2, align 4
+; CHECK-NEXT:  ret i32 %6
+; CHECK-NEXT:}
+
+;--------------------
+;|      TEST 5      |
+;--------------------
+
+define dso_local void @_Z15void_calls_voidv() #0 {
+  call void @_Z11void_no_argv()
```
Comment (m-ly4) [inline llvm/test/lab2/pushkarev_ilya/test.ll:212]: Could you please add a test where both functions have a loop or if statement?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**